### PR TITLE
[#112] FreeBSD support

### PR DESCRIPTION
**Pull request by akhramov**
FreeBSD is a somewhat popular system among developers and it would be
nice to have it supported.
In order to support it, we have to patch the remoteprocess subcrate a
bit, as well as bump `proc-maps` and `read-process-memory`
crates version.

This change

[remoteprocess crate]

Adds FreeBSD support for

* process and thread locking (via ptrace)
* getting thread registers (via ptrace)
* getting process' executable and cwd (via libprocstat)
* read_process_memory contract has changed in the latest version, so
  OSX / Linux / Windows code has been adjusted accordingly.

The only thing that is left to be implemented is unwinder /
symbolificator. Therefore `py-spy` on FreeBSD does not support native
stack traces, yet.

[py-spy]

* Adds FreeBSD-specific versions of platform-dependent functions.
* Adds Cirrus CI configuration to run tests against FreeBSD
  platform. In order to set CI up, please refer to
  https://cirrus-ci.org/guide/quick-start/

### CODE REVIEW

**Comment by akhramov** [comment]
@benfred 
Regarding dependencies... 

I have changed `read-proc-memory` to point to my fork.  The fork incorporates your changes from luser/read-process-memory#6 and my from luser/read-process-memory#10. I would be happy to have a better solution.

Also, `proc-maps` currently resolves to a git branch containing changes from rbspy/proc-maps#3. I believe, you can review / merge this one.

**Comment by akhramov** [comment]
Alright, windows tests seem to fail. I'm looking into this.

**Comment by akhramov** [comment]
The tests are green!

File: remoteprocess/Cargo.toml [hunk 1]
```diff
@@ -2,19 +2,13 @@
 name = "remoteprocess"
 version = "0.1.0"
 authors = ["Ben Frederickson <github@benfrederickson.com>"]
-repository = "https://github.com/benfred/py-spy"
-homepage = "https://github.com/benfred/py-spy/tree/master/remoteprocess"
-description = "cross platform api for getting information on a running processes"
-readme = "README.md"
-license = "GPL-3.0"
-build="build.rs"
```
Comment (benfred) [inline remoteprocess/Cargo.toml]: This block shouldn't be removed

File: remoteprocess/src/freebsd/lock.rs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+use super::Error;
+use super::ptrace;
+
+#[derive(Debug)]
+pub struct ProcessLock {
+    pid: pid_t,
+}
+
+impl ProcessLock {
```
Comment (benfred) [inline remoteprocess/src/freebsd/lock.rs]: With linux - we're locking each thread individually, using multiple ptrace attach calls: https://github.com/benfred/py-spy/blob/1717b91d7fafc6910e1e9b76e3397a7d58e883dc/remoteprocess/src/linux/mod.rs#L61-L75

Does attaching to only the main thread pause all the child threads as well?

If not, can we share the linux ptrace locks w/ bsd here?

**Comment by benfred** [comment]
Thanks for this! Mostly looks good. Also left a minor comment on the proc-maps PR.

One other thing - I'm considering relicensing this project from GPL3 to MIT (I want to share a bunch of code here w/ rbspy - and need a compatible license for that, also a MIT license will make adoption at my work a little easier). Are you good with releasing your change under a MIT license?

File: remoteprocess/src/freebsd/lock.rs (hunk 2 above)
Comment (akhramov) [reply remoteprocess/src/freebsd/lock.rs]: > Does attaching to only the main thread pause all the child threads as well?

Yes, in FreeBSD, `PT_ATTACH` sends `SIGSTOP` to the whole process:
https://github.com/freebsd/freebsd/blob/7f19ffe6f947e40299a4be93b7cf320da2e17951/sys/kern/sys_process.c#L955


I test this behavior indirectly here:
https://github.com/benfred/py-spy/blob/d79a69a4f6f380a179944106126ce88265c4bab1/remoteprocess/src/freebsd/mod.rs#L254-L261

In  contrary, Linux `ptrace` operates solely on threads:

>  Therefore,  "tracee"  always  means  "(one)  thread", never "a (possibly multithreaded) process".  Ptrace commands are always sent to a specific tracee using  a  call  of  the form

So, BSD ptrace gives us more flexibility here.

File: remoteprocess/Cargo.toml (hunk 1 above)
Comment (akhramov) [reply remoteprocess/Cargo.toml]: Sure, my bad!

**Comment by akhramov** [comment]
Thanks for the review!

I'm fine with releasing my code under MIT license. Moreover, I'm in no rush, you can accept this contribution after the relicensing.

**Comment by akhramov** [comment]
Bumped proc-maps version. Also, I created a PR to your `read-process-memory` fork (benfred/read-process-memory#2).

I want you to keep control over `py-spy` dependencies. :)

**Comment by benfred** [comment]
thanks!

**Comment by drozdowsky** [comment]
thank you

**Pull request merged**